\documentclass[a4paper]{article}
\usepackage{fancyvrb}
\usepackage{pdfpages}
\usepackage{url}
\begin{document}

\newcommand\vitem[1][]{\SaveVerb[% to use verb in description
    aftersave={\item[\textnormal{\UseVerb[#1]{vsave}}]}]{vsave}}

\title{\Huge \texttt{Minia-GATB} --- Short manual}

\author{R. Chikhi \& G. Rizk \& E. Drezen \& D. Lavenier\\
        {\small{rayan.chikhi@ens-cachan.org}}}
\maketitle

\begin{abstract}
\noindent {\normalsize Minia is a software for ultra-low memory DNA sequence assembly. It takes as input a set of short genomic sequences (typically, data produced by the Illumina DNA sequencer). Its output is a set of contigs (assembled sequences), forming an approximation of the expected genome. Minia is based on a succinct representation of the de Bruijn graph. The computational resources required to run Minia are significantly lower than that of other assemblers.}
\end{abstract}

\tableofcontents

\section{Forewords}

Minia-GATB is the codename of a new version of Minia using the GATB library. We will drop the "-GATB" suffix and just refer to it as "Minia". 
In terms of features, not much has changed: Minia remains a contigs assembler using very low memory. Notable changes are:
\begin{itemize}
    \item Different command line format 
    \item Much faster $k$-mer counting step
    \item Bubbles are collapsed using the most covered path (coverage information is now available for each node in the graph)
\end{itemize}

\section{Installation}

Downlad the binary and run Minia using the comand \verb+./minia+.

To install Minia from the sources, just type: \\ \verb+mkdir build && cd build && cmake .. && make -j 4+ \\
Minia has been tested on Linux and MacOS systems.

\section{Parameters}

The basic usage is:\\

\begin{verbatim}
    ./minia -in [input file] -kmer-size [kmer size] \
            -abundance-min [abundance value] -out [prefix]
\end{verbatim}


An example command line is:\\


\verb+./minia -in reads.fastq -kmer-size 31 -abundance-min 3 -out minia_assembly_k31_m3+\\

The main parameters are:

\begin{enumerate}

    \item \verb+in+ -- the input file(s) (see Section 5 for inputting multiple files)

    \item \verb+kmer-size+  -- k-mer length (integer)

    \item \verb+abundance-min+ -- specifies how many times a $k$-mer must be seen in the reads to be considered correct (integer)

\item \verb+prefix+ -- any prefix string to store output contigs as well as temporary files for this assembly

\end{enumerate}

Minia now uses the Cascading Bloom filters improvement (http://arxiv.org/abs/1302.7278) by default, thanks to Gustavo Sacomoto for the implementation in Minia. Launch Minia with the \verb!-debloom original! option to revert to the original data structure.


\section{Explanation of parameters}
\begin{description}

\vitem+kmer-size+
The $k$-mer length is the length of the nodes in the de Bruijn graph. It strongly depends on the input dataset. We recommend that you use the KmerGenie software to automatically find the best $k$ for your dataset.

\vitem+abundance-min+
The \verb+abundance-min+ is used to remove erroneous, low-abundance $k$-mers. This parameter also strongly depends on the dataset. Any $k$-mer that appears strictly less than \emph{abundance-min} times in the reads will be discarded. A typical value is $3$. It is also called "coverage cut-off" in other software (e.g. Velvet).

KmerGenie has an experimental feature (in the output of the command, not in the HTML report) that reports an optimal coverage cut-off (synonymous to abundance) parameter that exactly corresponds to the abundance-min parameter of Minia.

\textbf{Setting \emph{abundance-min} to $1$ is not recommended}, as no erroneous $k$-mer will be discarded, which will likely result in a very large memory usage. In particular, if KmerGenie tells you to set abundance to 1, then still, set it to 2. If the dataset has high coverage, try larger values. 

\vitem+prefix+
The \verb+prefix+ parameter is any arbitrary file name prefix, for example, \verb+test_assembly+.

\end{description}

\section{Input}

\begin{description}
\item \emph{Larger $k$-mer lengths}

Minia supports arbitrary large $k$-mer lengths. To compile Minia from the source, to support $k$-mer lengths up to, say, 320, type this in the build folder:
\begin{verbatim}
rm -Rf CMake* && cmake -DKSIZE_LIST="32   64   96  128  160  192  224  256  320" .. && make -j 4
\end{verbatim}

The list of kmers should only contain multiples of 32. Intermediate values are used to create optimized code for k values that are shorter than the max. The last k value specifies the highest kmer size that can effectively be used in this compiled version of Minia. Apart from that, whatever k values are in this list do not affect the assembly quality in any way.

\item \emph{FASTA/FASTQ}

Minia assembles any type of Illumina reads, given in the FASTA or FASTQ format. Giving paired or mate-pairs reads as input is OK, but keep in mind that Minia won't use pairing information.
\item \emph{Multipe Files}

 Minia can assemble multiple input files. Just create a text file containing the list of read files, one file name per line, and pass this list as the first parameter of Minia (instead of a FASTA/FASTQ file). Therefore the parameter \verb+input_file+ can be either (i) the read file itself (FASTA/FASTQ/compressed), or (ii) a file containing a list of file names.

\item \emph{Line format}

 In FASTA files, each read can be split into multiple lines, whereas in FASTQ, each read sequence must be in a single line.

\item \emph{Gzip compression}

Minia can direclty read files compressed with gzip. Compressed files should end with '.gz'. Input files of different types can be mixed (i.e. gzipped or not, in FASTA or FASTQ)

\item \emph{Graph input}

Minia can take as input a graph in the HDF5 format, constructed using the dbgh5 program, or a different GATB tool, or from a previous run of Minia (useful when one wants to tweak assembly parameters). Use the \verb+-graph+ parameter to specify the graph, instead of the \verb+-in+ parameter.

\end{description}

\section{Output}

The output of Minia is a set of contigs in the FASTA format, in the file \verb+[prefix].contigs.fa+. 

\subsection*{Creating unitigs}

Minia supports the creation of unitigs. Use the following command line: 
\begin{verbatim}
-starter simple -no-length-cutoff -traversal unitig
\end{verbatim}

The \verb+-starter simple+ option is to turn off Minia's heuristics for starting node selection (which avoid starting inside a bubble).
Note that there is a known bug in the \verb+-starter simple+ behavior: some of the unitigs of length k won't be returned (technically: the branching nodes which are neighbors of branching nodes are not returned).

\section{Memory usage}

We estimate that the memory usage of Minia is roughly $1$ GB of RAM per gigabases in the target genome to assemble. It is independent of the coverage of the input dataset, provided that the \verb!abundance-min! parameter is correctly set. For example, a human genome was assembled in $5.7$ GB of RAM. This was using the original data structure; the current implementation relies on Cascading Bloom filters and uses $\approx 4$ GB. A better estimation of the memory usage can be found in the Minia article.

\section{Disk usage}

Minia writes large temporary files during the k-mer counting phase. These files are written in the working directory when you launched Minia. For better performance, run Minia on a local hard drive. 

\section{Influences}

This version of Minia could not exist without inspiration taken from these great pieces of software:
\begin{itemize}
    \item khmer \url{https://github.com/dib-lab/khmer}
    \item KMC2 \url{https://github.com/marekkokot/KMC}
    \item SPAdes \url{http://bioinf.spbau.ru/spades}
\end{itemize}
\end{document}

